rfcs/2020-11-06-4909-remap-template-fields.md

```diff
@@ -0,0 +1,142 @@
+```
+
+The actual message sent to Loki would be:
+
+```
+{"message1": "thing", "message2": "thong"}
+```
+
+With the label `key = bar-zab` attached to the message. The fields `foo` and 
+`buzz` that were used in defining the label have been removed from the message.
+
+With Remap this list becomes dynamic. So the script `if .foo { .bar } else 
+{ .baz }` means sometimes `.bar` is used and other times `.baz` is used. It is 
+not clear how we should proceed to determine which fields need to be removed
+from the message.
+
+
+## Internal Proposal
+
+As there are a number of complications with working out which fields are used
+in the output, we could only support this feature if the old style template
+syntax is used. If Remap is used, the user won't have the option to remove these
+fields from the message in the Loki sink.
+
+(This is the simplest option.)
+
+## Rationale
+
+The benefits of using Remap in the template fields are:
+
+ - One familiar syntax and function reference for Vector.
+ - Access to all of remap's functions for templating.
+ - Less code to manage (once the old template fields are deprecated).
+
+
+## Drawbacks
+
+There could be an additional maintenance burden. Should the need to track fields
+used in the script be implemented that is a fairly significant complication to
+the script execution process that will need to always be kept in mind when 
+future additions to the language are being implemented.
+
+
+## Alternatives
+
+### Do nothing
+
+We do already have the existing template syntax. Perhaps we can stick with this.
+
+The advantage of using Remap for these fields are that it allows more 
+flexibility in defining how the event is used. However, given that remap can be 
+used as a transform, should the user really need this they could put a Remap 
+transform in the process to process these fields so they can be easily used in 
+the template for the next phase.
+
+### Returning fields at load time
+
+To fix the issue with returning fields so that the Loki sink removes them, whilst
+Remap is parsing the script, it could keep track of all fields that are being
+used in the script.
+
+This script, `if .foo { .bar } else { .baz }`, would result in all three
+fields being returned - `.foo`, `.bar` and `.baz`, and subsequently removed
+from the message sent to Loki.
+
+
+### Returning fields at run time
+
+Remap could keep track of all the fields that are read from as it runs and
+return these fields. So the script, `if .foo { .bar } else { .baz }`,
+would result in `.foo` and either `.bar` or `.baz` being returned.
+
+If necessary, Remap could distinguish between fields that are used in the
+condition and those used it the result, so only `.bar` or `.baz` could be
+returned.
+
+There are likely to be a number of edge cases that would need to be thought
+through if we took this approach. For example, if a field is used in a function
+but it's value is only used to influence the result - should it be included?
```

Comment:
Yeah, this seems like a no-go, again because there are way too many edge-cases and the added cognitive complexity it adds to the language.